### PR TITLE
Add gym levels to markers (alternative to #351) and misc gym marker upgrades

### PR DESCRIPTION
**Pull request by Studio-XP**
## Description
Alternative to #351, since it was closed by creator and never reopened. Adds gym levels to the gym markers on the map.

Restyled shield version of the markers incorporates a colored border that matches the team. Also created a second set of markers to match the in-game version. Selectable from Style Settings in the options

Upgrades include:
* Standardized, artifact free, compressed team logos for info windows
* High resolution gym markers so mobile users don't have to see a blurry version

## Motivation and Context
I liked the idea behind #351 and was sad that it was closed by the author and never reopened :(. Refactored the code to make it more reusable

## Screenshots (if appropriate):
782 - settings

In-Game Markers
782 - new pokegyms
Shield Markers
782 - gyms_shield

Mobile low-res vs high-res on phone (they are not this large on desktop!)
ingame blurry
ingame crisp
shield blurry
shield crisp

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: static/js/map.js [hunk 1]
```diff
@@ -1183,7 +1189,7 @@ function setupGymMarker (item) {
       lng: item['longitude']
     },
     map: map,
-    icon: 'static/forts/' + gymTypes[item['team_id']] + '.png'
+    icon: 'static/forts/' + gymTypes[item['team_id']] + (item['team_id'] === 0 ? '' : '_' + getGymLevel(item['gym_points'])) + '.png'
```
Comment (Zerotask) [inline static/js/map.js]: it's better to use the more frequent case: item['team_id'] !== 0 ...

File: static/js/map.js [hunk 2]
```diff
@@ -1196,7 +1202,7 @@ function setupGymMarker (item) {
 }
 
 function updateGymMarker (item, marker) {
-  marker.setIcon('static/forts/' + gymTypes[item['team_id']] + '.png')
+  marker.setIcon('static/forts/' + gymTypes[item['team_id']] + (item['team_id'] === 0 ? '' : '_' + getGymLevel(item['gym_points'])) + '.png')
```
Comment (Zerotask) [inline static/js/map.js]: same here.

File: static/js/map.js (hunk 1 above)
Comment (xrobau) [inline static/js/map.js]: When using ternary operators, its traditional to put the shortest use case first, for readability.  Not saying either way is right, but that's the way I do it.

**Comment by jnovack** [comment]
I absolutely, unequivocally ABHOR those icons, not your fault.  A map past zoom level 13 is completely unreadable.

Now's the time to change them to something more gooder.  You are welcome.

image

image 

image

image

image

**Comment by slideup-benni** [comment]
Dou you think this will be compatible with https://github.com/PokemonGoMap/PokemonGo-Map/pull/666?

**Comment by Studio-XP** [comment]
@slideup-benni Yes, I don't see why it wouldn't be perfectly compatible with #666. In fact, afaik, that more or less modifies the contents of the infoWindow, whereas this modifies the marker.

Looking at his code, the lines that I've modified are left untouched by his code, with the exception of his line 1086 on map.js, where he references gymPrestige. I moved the array that contains the values he's calling to within getGymLevel() so it's a local variable, but after looking at the code, I'm going to relocate that so it's a global variable under `var gymTypes`

**Comment by slideup-benni** [comment]
@Studio-XP Thank you for your detailed check :)

**Comment by Studio-XP** [comment]
@jnovack Honestly, anything at zoom level 13 is unreadable, such as pokemon, pokestops and gyms.  In fact, things start to get difficult at zoom level 14 in a suburb (I wouldn't want to imagine what a city looks like zoomed out). At least in my case, the gym markers are the only thing that are legible at that zoom level. I'm not sure what benefit changing the gym marker icons would have.

My map at zoom level 13, with pidgeys and rattatas filtered out
782 - zoom_level_13

**Comment by jnovack** [comment]
image

This is SO much easier to put in context than 50% white images.

**Comment by Studio-XP** [comment]
As a happy medium, what if we change/add a border around the current gym markers to their respected teams? That way it's not a whole mess of white with no info until the middle of the icon?

Off topic: we're not that bad, unless you're mystic ofc. I wish we had that many gyms in Delaware. Even in Wilmington, they're spread out

**Comment by Zerotask** [comment]
For the yellow icons ... 1 -5 is enough.

**Comment by Studio-XP** [comment]
@jnovack, here's a proof of concept for a slightly restyled gym marker.
forts_wip

**Comment by jnovack** [comment]
I appreciate the restyling, but is there an aversion to the full color icon?

The "shield" takes up quite a bit of pixel space.  In example 3, you cannot tell there's an unclaimed gym or third yellow gym (I know, impossible anyway).  The full color icons I've provided have a huge amount of transparency and a very long coastline for maximum pixelage.

Stylistically, the team logo (on or off a shield, does not matter) is extremely out of place on a map, especially when the unclaimed gym is a GYM icon, representing a gym.

**Comment by Studio-XP** [comment]
Are you against the idea of having gym levels on the markers too or are you ok with that? Based off of what it sounds like, you're against the whole idea of this pull request :/

**Comment by jnovack** [comment]
No, I LOVE the levels.  I HATE the shield icons/marker.

That's why I suggested a different marker.  I never said ANYTHING about the levels.

**Comment by Studio-XP** [comment]
Ok, so I'm gonna mess with this more. I've redone the gyms the way you prefer, still haven't found an aesthetically pleasing location for the levels yet, but here's a preview. Might make an option to choose between the two :/

782 - jnovack version

**Comment by jnovack** [comment]
https://dribbble.com/shots/2831980-Pok-mon-GO-Team-Logos-Vector-Download

The bonus "gym" logo is in the file.  EPS and SVG for your rendering heart's content.

**Comment by Studio-XP** [comment]
Is there something wrong with the in-game one's I've created that you don't like?

**Comment by jnovack** [comment]
There's nothing wrong with the "left" picture.  If you wanted an SVG vector logo, that's where I downloaded it from.  I was merely trying to help.

I have a big +1 for the left side icons.

**Comment by Studio-XP** [comment]
Haha, ok. Yeah, I just created a path of the large uncontested marker and went from there. Btw, sorry if I came across as being hostile earlier, long week... Tgif

**Comment by jnovack** [comment]
We're cool. :)

Perhaps a front-end change to toggle both icons and levels.

Don't forget to set defaults to use the awesome icons and levels. :)

image

**Comment by FlarumHostingCo** [comment]
+1 for the left ones!  Those look stunning!

**Comment by Studio-XP** [comment]
I'm gonna add support to change the gym marker style sometime tomorrow. I've got it done, I just have to add it here

**Comment by FlarumHostingCo** [comment]
@Studio-XP if you could just upload the left images to use with your old version now..  That be very cool  :)

**Comment by slideup-benni** [comment]
@jnovack @Studio-XP  but put the front-end changer in the style tab, please. Maybe with a dropdown for the gym-icons, so others can add more later

**Comment by Studio-XP** [comment]
@slideup-benni That was my plan all along.  Right under the Search Icon Marker (which tbh should be named Search Marker Icon)

**Comment by jnovack** [comment]
You will only get my blessing if the In-Game icons are by default. ;)

**Comment by Studio-XP** [comment]
I don't want to confuse the general public or get them upset like the commit about changing the default search marker to "Mario" haha

**Comment by Becold** [comment]
+1 Tested and looks ok

**Comment by jnovack** [comment]
This has a 1 conflict (from #666) that can EASILY be resolved.

I've been using this for days (on the in-game icons of course), and it's beautiful.

**Comment by jnovack** [comment]
@Studio-XP Fix the merge conflict and this has my blessing.  LGTM!

**Comment by FlarumHostingCo** [comment]
What is the conflict?  It works fine for me.

**Comment by jnovack** [comment]
Did you miss the HUGE gray icon and the bold statement saying "This branch has conflicts that must be resolved" immediately above the input box where you typed?

**Comment by Studio-XP** [comment]
Yeah, I've gotta rebase it when I get home

**Comment by jnovack** [comment]
> Yeah, I've gotta rebase it when I get home. I haven't had time to look at his yet, but I would imagine it's about the gymlevel?

Yeah, it's a single line in map.js.

**Comment by Studio-XP** [comment]
Holy shit, I suck at Git... took me forever to figure out how to rebase and fix conflicts :(

**Comment by jnovack** [comment]
Oh, no no no.. Git Desktop is not how you want to merge conflicts.

```
git checkout develop
git pull
git checkout patch-1  // Your local branch name
git rebase -i develop
// Now, go and physically inspect the files.
git rebase --continue
```

`rebase -i`  will pause the rebase after all the conflicts but before the final merge.  If there are conflicts, you are in a "detached HEAD" state, sort of like a floating ghost state.  Go physically inspect all the files it complained about.  Search for "<<<<<<" and ">>>>>>".  LessThan means theirs, GreaterThan means yours.  It couldn't figure out how to do it automatically (that's a good thing).

Remove all the `<`, `>` and `=` lines and make the section as it should be.  Then do the `rebase --continue`.

**Comment by Studio-XP** [comment]
That's how I essentially did it. For some reason my linux box doesn't have git setup properly to push commits, so I had to do it with windows :/

**Comment by Studio-XP** [comment]
@jnovack Finally pulled the trigger and changed the default gym marker to the more visually appealing in-game markers :)

**Comment by jnovack** [comment]
I'm glad you see the value. :)

**Comment by djv2** [comment]
can this be merged yet :) ?

**Comment by Studio-XP** [comment]
Thanks hatter. Honestly haven't put any time into this supporting this project for 2 weeks. Didn't even realize it had conflicts :)

**Comment by SoonDead** [comment]
THIS is what the develop branch needs

**Comment by jnovack** [comment]
@maddhatter AWWWW SHIT.. Approving Changes. <3

**Comment by maddhatter** [comment]
Just one more way of jumping up and down, going, "Please, for the love of God - merge this!"

**Comment by CalcMan** [comment]
Just testing this out, the icons are larger then the old ones, and they take up too much screen realestate.

The map (with gyms turned on) has always been unusable at high zoom levels, but with these larger icons sizes we reach the non-usable level at a lower zoom then we had earlier.

So 3 items:
1 - can we make icons scale with zoom level (we haven't done it before)
2 - can the images be made smaller (to be the same size as the old icons)
3 - don't remove the old icons (those original ones w/o numbers), give people the choice

**Comment by Studio-XP** [comment]
Are you referencing what you're currently seeing on your map or just what I've uploaded?

The code `scaledSize: new google.maps.Size(48, 48)` in map.js should take care of scaling them back to a smaller size (48x48)

**Comment by CalcMan** [comment]
I just pulled your code base into my running map, so I'm referring to what I am seeing. The new shields are slightly larger (they have the color border), and the new ingame style icons are bigger then the new shields.

Also, seems like I'm getting some exception ERRORS (there are many like this):

2016-09-25 16:12:05,761 [       Thread-74][           app][   ERROR] Exception on /static/forts/shield/Mystic_7.png [GET]
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/flask/app.py", line 1988, in wsgi_app
    response = self.full_dispatch_request()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/flask/app.py", line 1641, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/flask/app.py", line 1544, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/flask/app.py", line 1637, in full_dispatch_request
    rv = self.preprocess_request()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/flask/app.py", line 1837, in preprocess_request
    rv = func()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/playhouse/flask_utils.py", line 165, in connect_db
    self.database.connect()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/peewee.py", line 3402, in connect
    **self.connect_kwargs)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/playhouse/pool.py", line 127, in _connect
    raise ValueError('Exceeded maximum connections.')

**Comment by Studio-XP** [comment]
Could you try pulling this into a new clone to see if you get the same errors? You can also contact me on Discord for further help @xplozion

**Comment by jnovack** [comment]
> File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/playhouse/pool.py", line 127, in _connect raise ValueError('Exceeded maximum connections.')

This is NOT a PR error. This is a Flask problem.

#LGTM

**Comment by Studio-XP** [comment]
@primaxius Yes sir, reverted the default back to the shields.

Thank you

**Comment by marumari** [comment]
@Studio-XP:

I really love this feature, but can I add a request?  It would be nice to have a simple ⭐ icon (or something) on a gym icon, if you had a Pokémon in the gym.

You'd simply have to enter your trainer name into the details, and it would give people the most useful piece of information about gym status without having to do anything complicated like a gym details sidebar.

**Comment by Studio-XP** [comment]
@marumari I'm glad you got your marker size issue sorted out. It looked like you were running an old version of map.js. I'm assuming you ran npm build/grunt build after downloading from the head originally, correct? Either way, the code in map.js wasn't loaded into your local copy. If it was, it would have shown the gym levels on the icons

On to your request:
So markers work by loading a static image as the marker. The Google Map api doen't allow multiple images to be set to a single marker. So in order to accomplish this, we would have to create a new set of dedicated gym markers for the possibility that you own said gym at a specific level. Essentially doubling the amount of gym markers in the forts directories. So to implement your idea, we would go from 60 gym level icons (30 x 2 sets) to 120 icons, and if we also implement the @CalcMan's comment from #1407, it would double again to 240. Now I doubt this project follows the KISS principle, but in the offshoot that it does, I believe this would go highly against it :)

**Comment by marumari** [comment]
Have you seen the notification code? I wouldn't call it KISS. :P

Is it safe to assume that the black circles are glyphs and not done via CSS? Otherwise, I'd say that the circles could be green instead of black. Maybe not that much less complex with regards to the number of icons, but at least easier to implement.

**Comment by Studio-XP** [comment]
Yes, the levels are in the images themselves. It was a painstaking task just to make the 60 images in this commit :/

**Comment by marumari** [comment]
Were they created in Photoshop? It should be pretty simple to automate their creation if you have the PSD files.

**Comment by CalcMan** [comment]
I wonder if there is a way to use multiple markers for each gym (at the same latlon), and use z-order to put the extra details on top.

Reference: https://developers.google.com/maps/documentation/javascript/markers

**Comment by Studio-XP** [comment]
Yes, I made them in PS and still have the psd just in case anyone wants to contribute to the gym markers, but to hide one layer and display the next layer for a new gym marker, hit save, rename, etc... :) After a minute I was wanting it to be over. The best was when I finished and forgot to enable a layer or a style, thus forcing me to start all over :(

**Comment by marumari** [comment]
Woah, that's crazy! There should just be seven layers: two for each team and one black circle. Eight if you add a green circle.

The numbers, saving, etc. could all just be a photoshop script. The whole process shouldn't take more than a minute or so once the script is created.

**Pull request merged**